### PR TITLE
Cherry pick basic DA functionality from celestia-core

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

Mainly cherry picks #83, with a few modifications to better fit in a newer version of tendermint, and this PR doesn't add the data availability header to the block. 

part of #528

### CODE REVIEW

File: internal/mempool/v0/clist_mempool_test.go [hunk 1]
```diff
@@ -121,11 +121,11 @@ func TestReapMaxBytesMaxGas(t *testing.T) {
...
 		{20, 10, 10, 0},
-		{20, 24, 10, 1},
+		{20, 28, 10, 1}, // account for overhead in Data{}
 		{20, 240, 5, 5},
-		{20, 240, -1, 10},
-		{20, 240, 10, 10},
-		{20, 240, 15, 10},
+		{20, 280, -1, 10},
+		{20, 280, 10, 10},
+		{20, 280, 15, 10},
```
Comment (evan-forbes) [inline internal/mempool/v0/clist_mempool_test.go:128]: this is due to the increased size of proto `Data`

see this func https://github.com/celestiaorg/celestia-core/blob/8ea76d0823d56b68504b80fa394cfa6f39e5dfce/types/tx.go#L154-L160

File: types/block.go [hunk 2]
```diff
@@ -107,6 +109,73 @@ func (b *Block) fillHeader() {
...
+// // fillDataAvailabilityHeader fills in any remaining DataAvailabilityHeader fields
+// // that are a function of the block data.
+// func (b *Block) fillDataAvailabilityHeader() {
+// 	namespacedShares := b.Data.computeShares()
+// 	shares := namespacedShares.RawShares()
+// 	if len(shares) == 0 {
+// 		// no shares -> no row/colum roots -> hash(empty)
+// 		b.DataHash = b.DataAvailabilityHeader.Hash()
+// 		return
+// 	}
```
Comment (evan-forbes) [inline types/block.go:121]: planning to do this in a different PR, as share splitting and merging isn't added yet

File: types/block.go [hunk 3]
```diff
@@ -44,8 +47,7 @@ type Block struct {
 
 	Header     `json:"header"`
 	Data       `json:"data"`
-	Evidence   EvidenceData `json:"evidence"`
-	LastCommit *Commit      `json:"last_commit"`
+	LastCommit *Commit `json:"last_commit"`
```
Comment (evan-forbes) [inline types/block.go:50]: note that the data availability header is not added to the block. Instead, we can simply generate the DAH and data hash, then discard it. 

If we want to save it for whatever reason, then we can add it seperately to `BlockMeta`

ref #512

File: types/block_test.go [hunk 4]
```diff
@@ -470,10 +465,10 @@ func TestBlockMaxDataBytes(t *testing.T) {
...
 		1: {10, 1, 0, true, 0},
 		2: {841, 1, 0, true, 0},
-		3: {842, 1, 0, false, 0},
-		4: {843, 1, 0, false, 1},
-		5: {954, 2, 0, false, 1},
-		6: {1053, 2, 100, false, 0},
+		3: {845, 1, 0, false, 0},
+		4: {846, 1, 0, false, 1},
+		5: {956, 2, 0, false, 1},
+		6: {1056, 2, 100, false, 0},
```
Comment (evan-forbes) [inline types/block_test.go:471]: increased because of the fields added to `Data`

**Review by liamsi — APPROVED** [review]
👍🏼
